### PR TITLE
Add unit test for ContinuousBuildActionExecutor

### PR DESCRIPTION
**Pull request by wolfs**
Re-enables and extends the unit test for `ContinuousBuildActionExecutor`.

### CODE REVIEW

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 1]
```diff
@@ -78,8 +96,6 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
 
         then:
         1 * delegate.execute(action, buildSessionContext)
-        1 * deploymentRegistry.runningDeployments >> []
-        0 * waiterFactory._
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:82]: Add `0 * _`.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 2]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+        cleanup:
+        println(textOutputFactory)
     }
 
     def "exits if there are no file system inputs"() {
         when:
         continuousBuild()
         1 * delegate.execute(action, buildSessionContext)
+        then:
         executeBuild()
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:208]: Check console output.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 3]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+        then:
+        buildExits()
+
+        cleanup:
+        println(textOutputFactory)
     }
 
     def "exits if there are no file system inputs"() {
         when:
         continuousBuild()
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:205]: Call this `continousBuildWithoutInputs()`?

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 4]
```diff
@@ -55,20 +61,32 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+    def deploymentRegistry = Stub(DeploymentRegistryInternal) {
+        runningDeployments >> deployments
+        executionGate >> continuousExecutionGate
+    }
     def buildSessionContext = Mock(BuildSessionContext)
+    def textOutputFactory = new TestStyledTextOutputFactory()
+    def executorService = Executors.newCachedThreadPool()
+
     def executer = executer()
+    Future<?> runningBuild
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:79]: Pass this as parameter instead?

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 5]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
         System.in.read() == -1
     }
 
-    def "waits for waiter"() {
+    def "can cancel the build"() {
         when:
-        continuousBuild()
-        1 * delegate.execute(action, buildSessionContext) >> {
-            declareInput(file)
+        continuousBuildWithInputs()
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:128]: Use a `given:` block for this.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 6]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
-        1 * waiter.wait(_, _) >> {
-            cancellationToken.cancel()
+        conditions.eventually {
+            rebuiltBecauseOfChange()
         }
+
+        when:
+        cancellationToken.cancel()
+        then:
+        buildExits()
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:165]: Move this to `cleanup:`? Remove it maybe?

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 7]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+            rebuiltBecauseOfChange()
+        }
+
+        when:
+        cancellationToken.cancel()
+        then:
+        buildExits()
+
+        cleanup:
+        println(textOutputFactory)
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:200]: Remove debug stuff.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 8]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+
+        when:
+        cancellationToken.cancel()
+        then:
+        buildExits()
+    }
+
+    def "only triggers the build when the gatekeeper is open"() {
+        when:
+        buildWithDeployment()
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:170]: Use `given:`.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 9]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+        conditions.eventually {
+            reloadableDeploymentDetected()
+        }
+
+        when:
+        changeListeners.broadcastChange(FileWatcherRegistry.Type.MODIFIED, file.toPath())
+        // Wait to make sure the build is not triggered
+        Thread.sleep(500)
+        then:
+        waitingForChanges()
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:184]: Move up to after line 176.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 10]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+        runningBuild = executorService.submit {
+            executeBuild()
+        }
+        return runningBuild
+    }
+
+    private void continuousBuildWithInputs() {
+        buildWithInputs()
+        continuousBuild()
+    }
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:225]: Inline this.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:82]: Added.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 4 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:79]: Done.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 11]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
         when:
-        continuousBuild()
-        1 * delegate.execute(action, buildSessionContext) >> {
-            declareInput(file)
+        continuousBuildWithInputs()
+        runningBuild = startBuild()
+
+        then:
+        conditions.eventually {
+            lastLogLine == "Waiting for changes to input files of tasks..."
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:133]: I left a comment here so we don't forget: https://github.com/gradle/gradle/issues/18746#issuecomment-1023215551

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 5 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:128]: Done.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 12]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+    }
+
+    def "runs a new build on file change"() {
+        when:
+        continuousBuildWithInputs()
+        startBuild()
+
+        then:
+        conditions.eventually {
+            waitingForChanges()
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:151]: Renamed.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 6 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:165]: Moved to `cleanup:`.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 7 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:200]: Removed.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 8 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:170]: Done.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 9 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:184]: Done.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 13]
```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
...
+        waitingForChanges()
+        !changeDetected()
+
+        when:
+        gatekeeper.open()
+        then:
+        conditions.eventually {
+            rebuiltBecauseOfChange()
+        }
+
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:193]: Added.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 10 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:225]: Done.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy [hunk 14]
```diff
@@ -86,24 +87,31 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
     }
 
     def cleanup() {
+        println("Build Output:")
+        println(textOutputFactory)
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:91]: Do we want to keep these?

**Review by lptr — APPROVED** [review]
LGTM, I think it reads pretty nicely now.

File: subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy (hunk 14 above)
Comment (wolfs) [reply subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy:91]: Yeah, I added those so you see the build output in each test case. I think it helps in understanding that the test does the correct thing when you actually run them.

**Comment by wolfs** [comment]
@bot-gradle test and merge.

**Pull request merged**